topdown/http_test.go

```diff
@@ -2514,3 +2518,72 @@ var httpSendHelperRules = []string{
 		no_headers = object.remove(resp, ["headers"])
 	}`,
 }
+
+func httpSocket(path string) (net.Listener, error) {
+	sock, err := net.Listen("unix", path)
+	if err != nil {
+		return nil, err
+	}
+
+	// install signal handler to clean up
+	sigc := make(chan os.Signal, 1)
+	signal.Notify(sigc, os.Interrupt, syscall.SIGTERM)
+	go func(c chan os.Signal) {
+		<-c
+		sock.Close()
+		os.Exit(0)
+	}(sigc)
+
+	return sock, nil
+}
+func TestHTTPWithSocket(t *testing.T) {
+	tmpF, err := ioutil.TempFile("", "")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	socketPath := tmpF.Name()
+	tmpF.Close()
+	os.Remove(socketPath)
+
+	socket, err := httpSocket((socketPath))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	rs := http.Server{
+		Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.WriteHeader(http.StatusOK)
+		}),
+	}
+
+	go func() {
+		_ = rs.Serve(socket)
+	}()
+	defer rs.Close()
+
+	path := fmt.Sprintf("socket=%s", url.PathEscape(socketPath))
+	rawURL := fmt.Sprintf("unix://localhost/end/point?%s", path) // Send a request to the server over the socket
+	ok, rawURL, tr := useSocket(rawURL, nil)
+	if !ok {
+		t.Fatal("useSocket() returned !ok")
+	}
+
+	client := http.Client{
+		Transport: tr,
+	}
+
+	req, err := http.NewRequest("GET", rawURL, nil)
```

Comment:
I can see the value in this -- we're verifying that the transport is constructed properly -- but it's also testing something underneath the builtin. What do you think, would a test using `http.send` with a socket be helpful?

Reply:
I think that calling the built in is an excellent idea.  I found an example in [rego_test,go](https://github.com/open-policy-agent/opa/blob/bac175533bf5c1f902bbc513325a0554a9cd34a4/rego/rego_test.go#L1859).  I initially copied that test to http.go, but I ran into a recursive import.  rego_test.go imports [topdown](https://github.com/open-policy-agent/opa/blob/bac175533bf5c1f902bbc513325a0554a9cd34a4/rego/rego_test.go#L30) where my updates live.

Should a native rego socket test live there?  The current functions I added are not exported.

Reply:
There are a few test cases using the builtin in _this file_, too, I believe, see https://github.com/open-policy-agent/opa/blob/2d58e8b3ada39d471f79afa44b31ffb8f3825ea5/topdown/http_test.go#L44-L100

Reply:
@srenatus I copied portions of that test and replaced the existing one.  In [useSocket() ](https://github.com/open-policy-agent/opa/pull/3667/files#diff-4a137f284e9009d157ac4553842cb1c6e9260173fa29d3ed0d779f79e5b7a316R260)tlsConfig is always  set. I had initially assumed that it would be nil if the request was http and I was returning the URL as either http OR https base on its existence.  

Is there a scenario were https & socket go together?